### ISSUE TITLE
Datatable liveScroll broken by sorting

### ISSUE THREAD

**Issue by ohlsen3000** [primefaces/primefaces#1321]
In Primefaces 5.3 / Chrome:

```
<p:dataTable id="cars" widgetVar="carTable" var="car" rowKey="#{car.id}" 
    value="#{myView.lazyCarModel}" emptyMessage="#{msg['i18n.NO_ENTRY']}" 
    selectionMode="single" selection="#{myView.selectedCar}" 
    rows="15" scrollRows="15" 
    lazy="true" scrollable="true" liveScroll="true" scrollHeight="500">

   <p:column headerText="#{msg['i18n.car.ID']}" sortBy="#{car.id}" width="70">
        <h:outputText value="#{car.id}"/>
   </p:column>

</p:dataTable>
```
### HOW TO REPRODUCE
1. Open page with table
2. Perform live scrolling (at least up to 2nd page)
3. Sort by any column
4. Try to liveScroll > live scrolling will (in most cases) only load the first page, `load` method of `lazyCarModel` is not called anymore.

**Comment by ohlsen3000** [comment]
Need to call the following js on `sort` event to get things working again:

```
PF(widgetId).scrollBody.scrollTop(1);
PF(widgetId).allLoadedLiveScroll = false;
PF(widgetId).scrollOffset = 0;
```

**Comment by mvitorovic** [comment]
Can the developers please try and fix this for 6.0, so that we have lazy loading + live scrolling + filtering taken care of? At least the two most visible bugs in 5.3?

**Comment by kukel** [comment]
Did you try a 6.0 RC? there have been many direct fixes in the mean time and many indirect... And many combinations of the datatable that do not work are not directly bugs... Many combinations are just not tested (there is no multi-dimensional 'support matrix' for the datatable)

**Comment by mvitorovic** [comment]
I tested with RC2, this is what happens without the workaround:

Opened the page scrolled down:

2016-05-06 08:31:15,411 FINEST [*.AuditLazyModel](default task-51) ---->pageSize: 50
2016-05-06 08:31:15,411 FINEST [*.AuditLazyModel](default task-51) ---->first: 0
2016-05-06 08:31:26,110 FINEST [*.AuditLazyModel](default task-52) ---->pageSize: 50
2016-05-06 08:31:26,111 FINEST [*.AuditLazyModel](default task-52) ---->first: 50
2016-05-06 08:31:31,313 FINEST [*.AuditLazyModel](default task-53) ---->pageSize: 50
2016-05-06 08:31:31,313 FINEST [*.AuditLazyModel](default task-53) ---->first: 100

Selected a column for sorting

2016-05-06 08:31:34,170 FINEST [*.AuditLazyModel](default task-54) ---->pageSize: 50
2016-05-06 08:31:34,171 FINEST [*.AuditLazyModel](default task-54) ---->first: 0

Scrolled down:
2016-05-06 08:31:34,269 FINEST [*.AuditLazyModel](default task-55) ---->pageSize: 50
2016-05-06 08:31:34,269 FINEST [*.AuditLazyModel](default task-55) ---->first: 150  <---- Should not continue at 150

Switched the sorting column again

2016-05-06 08:31:50,082 FINEST [*.AuditLazyModel](default task-56) ---->pageSize: 50
2016-05-06 08:31:50,082 FINEST [*.AuditLazyModel](default task-56) ---->first: 0

Scrolled down again:
2016-05-06 08:31:59,469 FINEST [*.AuditLazyModel](default task-57) ---->pageSize: 50
2016-05-06 08:31:59,469 FINEST [*.AuditLazyModel](default task-57) ---->first: 200   <---- should not continue at 200

**Comment by hink084** [comment]
Similar in nature, I also found there to be a bug with Filtering/LiveScrolling with Primefaces v6.0.  @ohlsen3000 's solution works for Sorting but it does not with Filtering.  You can reproduce this by adding the "filterBy" attribute to the "p:column" tag.  In @ohlsen3000  's example this would be `filterBy="#{car.id}"`

The problem occurs when you attempt to scroll, filter/reduce the dataset, and then remove your filter to restore the dataset back to its original state - not necessarily in that order.

The following cases were tested:
1. Filter --> Remove Filter --> Scroll
2. Filter --> Scroll --> Remove Filter --> Scroll
3. Scroll --> Filter --> Remove Filter --> Scroll
4. Scroll --> Filter --> Scroll --> Remove Filter --> Scroll
5. Filter --> Filter --> {Empty Dataset} --> Remove Filters --> Scroll
6. Scroll --> Filter --> Filter --> {Empty Dataset} --> Remove Filters --> Scroll

I think there is a problem with the logic in the file: _/src/main/resources/META-INF/resources/primefaces/datatable/datatable.js_.  In particular the variables: _allLoadedLiveScroll_, _shouldLiveScroll_.  

You can see here in the loadLiveRows() function the _shouldLiveScroll_ variable gets set to false when there is no more data left to scroll (which can occur when filtering to empty datasets); this variable is never reset to true.

```
/**
 * Loads rows on-the-fly when scrolling live
 */
 loadLiveRows: function() {
     ...

     //Disable scroll if there is no more data left
     if(this.scrollOffset === this.cfg.scrollLimit) {
         this.shouldLiveScroll = false;
     }

     ...
}
```

The _allLoadedLiveScroll_ variable is set in this conditional statement:
```
if(this.cfg.liveScroll) {
    this.scrollOffset = 0;
    this.cfg.liveScrollBuffer = (100 - this.cfg.liveScrollBuffer) / 100;
    this.shouldLiveScroll = true;       
    this.loadingLiveScroll = false;
    this.allLoadedLiveScroll = $this.cfg.scrollStep >= $this.cfg.scrollLimit;      
}
```

and is adjusted in the liveScroll() oncomplete function
```
oncomplete: function(xhr, status, args) {
    if(args.totalRecords) {
        $this.cfg.scrollLimit = args.totalRecords;
    }

    $this.loadingLiveScroll = false;
    $this.allLoadedLiveScroll = ($this.scrollOffset + $this.cfg.scrollStep) >= $this.cfg.scrollLimit;
}
```
 but never adjusted to handle filtered rows or sorting.



As a quick fix for both Sorting and Filtering while attempting to LiveScroll I implemented this solution:

**1. Created a JS Object:**

```
var bugs = {
        /**
         * @param Object obj The Primefaces Datatable Object; Example: PF('widget_id')
         * @param Integer val The total rows available in the Datatable. When filtering this value will change.
         */
        fix_primefaces_livescroll : function(obj, val){
            // Reset the Scroll Position
            obj.scrollOffset = 0;
            obj.scrollBody.scrollTop(0);

            // Reset the Ability to Livescroll
            if(obj.cfg.scrollStep < val){
                obj.allLoadedLiveScroll = false;
                obj.shouldLiveScroll = true;
            }
        }
    }
};
```

**2. Attached the following Primefaces AJAX Events to the DataTable:**

```
<p:ajax event="sort"
        oncomplete="bugs.fix_primefaces_livescroll(PF('wgt_var'), args.totalRecords);"/>

<p:ajax event="filter"
        oncomplete="bugs.fix_primefaces_livescroll(PF('wgt_var'), args.totalRecords);"/>
```

**Comment by hink084** [comment]
This filtering bug appears to have been fixed in #2087 via [80df8cf0674b585e2cb3743928ff30594b75e411] by @mertsincan.  However, the sorting issue is still at large.  I created pull request: #2092
